Fix vagrant

treydock opened the pull request:
# Pull Request Checklist

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Fixes from #1233 to use correct version of stdlib as well as fix issue with SSL certs that get used.

ghoneycutt commented:
Need fixes for
- [ ] debian 8
- [ ] ubuntu 1604

ghoneycutt commented:
**Debian 8**

```
==> debian8-agent: Checking for guest additions in VM...
    debian8-agent: No guest additions were detected on the base box for this VM! Guest
    debian8-agent: additions are required for forwarded ports, shared folders, host only
    debian8-agent: networking, and more. If SSH fails on this machine, please install
    debian8-agent: the guest additions and repackage the box to continue.
    debian8-agent:
    debian8-agent: This is not an error message; everything may continue to work properly,
    debian8-agent: in which case you may ignore this message.
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

apt-get install -y linux-headers-`uname -r` dkms

Stdout from the command:

Reading package lists...
Building dependency tree...
Reading state information...


Stderr from the command:

E: Unable to locate package linux-headers-3.16.0-9-amd64
E: Couldn't find any package by regex 'linux-headers-3.16.0-9-amd64'
```

ghoneycutt commented:
**Ubuntu 1604**

```
==> ubuntu1604-agent: Running provisioner: shell...
    ubuntu1604-agent: Running: inline script
    ubuntu1604-agent: Error: Evaluation Error: Error while evaluating a Resource Statement, Unknown resource type: 'apt::source' (file: /etc/puppetlabs/code/environments/production/modules/sensu/manifests/repo.pp, line: 42, column: 5) on node ubuntu1604-agent.example.com
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

ghoneycutt commented:
**win2008r2**

```
==> win2008r2-agent: Machine booted and ready!
[win2008r2-agent] GuestAdditions versions on your host (6.0.14) and guest (4.3.12) do not match.
Copy iso file /Applications/VirtualBox.app/Contents/MacOS/VBoxGuestAdditions.iso into the box $env:TEMP/VBoxGuestAdditions.iso
^@^@^@The term 'Mount-DiskImage' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

At line:1 char:16
+ Mount-DiskImage <<<<  -ImagePath $env:TEMP/VBoxGuestAdditions.iso
    + CategoryInfo          : ObjectNotFound: (Mount-DiskImage:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
==> win2008r2-agent: Checking for guest additions in VM...
    win2008r2-agent: The guest additions on this VM do not match the installed version of
    win2008r2-agent: VirtualBox! In most cases this is fine, but in rare cases it can
    win2008r2-agent: prevent things such as shared folders from working properly. If you see
    win2008r2-agent: shared folder errors, please make sure the guest additions within the
    win2008r2-agent: virtual machine match the version of VirtualBox you have installed on
    win2008r2-agent: your host and reload your VM.
    win2008r2-agent:
    win2008r2-agent: Guest Additions Version: 4.3.12
    win2008r2-agent: VirtualBox Version: 6.0
The following WinRM command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

Mount-DiskImage -ImagePath $env:TEMP/VBoxGuestAdditions.iso

Stdout from the command:



Stderr from the command:

The term 'Mount-DiskImage' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

At line:1 char:16
+ Mount-DiskImage <<<<  -ImagePath $env:TEMP/VBoxGuestAdditions.iso
    + CategoryInfo          : ObjectNotFound: (Mount-DiskImage:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

ghoneycutt commented:
**Debian 9**

```
==> debian9-agent: Running provisioner: shell...
    debian9-agent: Running: inline script
    debian9-agent: Error: Evaluation Error: Error while evaluating a Resource Statement, Unknown resource type: 'apt::source' (file: /etc/puppetlabs/code/environments/production/modules/sensu/manifests/repo.pp, line: 42, column: 5) on node debian9-agent.example.com
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

ghoneycutt commented:
**Ubuntu 1804**

```
==> ubuntu1804-agent: Running provisioner: shell...
    ubuntu1804-agent: Running: inline script
    ubuntu1804-agent: Error: Evaluation Error: Error while evaluating a Resource Statement, Unknown resource type: 'apt::source' (file: /etc/puppetlabs/code/environments/production/modules/sensu/manifests/repo.pp, line: 42, column: 5) on node ubuntu1804-agent.example.com
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

ghoneycutt commented:
**EL 8**

```
==> el8-agent: Mounting shared folders...
    el8-agent: /vagrant => /Users/gh/git/sensu-puppet
Vagrant was unable to mount VirtualBox shared folders. This is usually
because the filesystem "vboxsf" is not available. This filesystem is
made available via the VirtualBox Guest Additions and kernel module.
Please verify that these guest additions are properly installed in the
guest. This is not a bug in Vagrant and is usually caused by a faulty
Vagrant box. For context, the command attempted was:

mount -t vboxsf -o uid=1000,gid=1000 vagrant /vagrant

The error output from the command was:

/sbin/mount.vboxsf: mounting failed with the error: No such device
```

ghoneycutt commented:
**Debian 10**

```
==> debian10-agent: Running provisioner: shell...
    debian10-agent: Running: inline script
    debian10-agent: Error: Evaluation Error: Error while evaluating a Resource Statement, Unknown resource type: 'apt::source' (file: /etc/puppetlabs/code/environments/production/modules/sensu/manifests/repo.pp, line: 42, column: 5) on node debian10-agent.example.com
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

ghoneycutt commented:
**win2012r2-agent**

```
==> win2012r2-agent: Running provisioner: shell...
    win2012r2-agent: Running: inline PowerShell script
    win2012r2-agent: Info: Loading facts
    win2012r2-agent: Info: Loading facts
    win2012r2-agent: Info: Loading facts
    win2012r2-agent: Info: Loading facts
    win2012r2-agent: Info: Loading facts
    win2012r2-agent: Notice: Compiled catalog for win2012r2-agent.attlocal.net in environment production in 2.48 seconds
    win2012r2-agent: Info: Applying configuration version '1585324250'
    win2012r2-agent: Info: Computing checksum on file C:/ProgramData/Sensu/config/agent.yml
    win2012r2-agent: Info: /Stage[main]/Sensu::Common/File[C:/ProgramData/Sensu/config/agent.yml]: Filebucketed C:/ProgramData/Sensu/config/agent.yml to puppet with sum 2e9fb201cb5485808f44e081e7399f69
    win2012r2-agent: Notice: /Stage[main]/Sensu::Common/File[C:/ProgramData/Sensu/config/agent.yml]/ensure: removed
    win2012r2-agent: Notice: /Stage[main]/Sensu::Cli/Archive[sensu-go-cli.zip]/ensure: download archive from https://s3-us-west-2.amazonaws.com/sensu.io/sensu-go/5.14.1/sensu-go_5.14.1_windows_amd64.zip to C:\Program Files\Sensu\sensu-go-cli.zip and extracted in C:\Program Files\Sensu without cleanup
    win2012r2-agent: Notice: /Stage[main]/Sensu::Cli/Windows_env[sensuctl-path]/ensure: created
    win2012r2-agent: Info: Class[Sensu::Cli]: Unscheduling all events on Class[Sensu::Cli]
    win2012r2-agent: Info: Stage[main]: Unscheduling all events on Stage[main]
    win2012r2-agent: Error: Could not set 'present' on ensure: The parameter is incorrect. - CreateProcess (file: C:/ProgramData/PuppetLabs/code/environments/production/modules/sensu/manifests/cli.pp, line: 89)
    win2012r2-agent: Error: Could not set 'present' on ensure: The parameter is incorrect. - CreateProcess (file: C:/ProgramData/PuppetLabs/code/environments/production/modules/sensu/manifests/cli.pp, line: 89)
    win2012r2-agent: Wrapped exception:
    win2012r2-agent: The parameter is incorrect. - CreateProcess
    win2012r2-agent: Error: /Stage[main]/Sensu::Cli/Sensuctl_configure[puppet]/ensure: change from 'absent' to 'present' failed: Could not set 'present' on ensure: The parameter is incorrect. - CreateProcess (file: C:/ProgramData/PuppetLabs/code/environments/production/modules/sensu/manifests/cli.pp, line: 89)
    win2012r2-agent: Notice: Applied catalog in 23.01 seconds
^@==> win2012r2-agent: Running provisioner: shell...
    win2012r2-agent: Running: inline PowerShell script
    win2012r2-agent: {
    win2012r2-agent:   version => "5.18.1",
    win2012r2-agent:   build => "c3f43c9d40f2b3fc34d32c6be0194e9557023035",
    win2012r2-agent:   built => "2020-03-10T18:32:19Z"
    win2012r2-agent: }
```

ghoneycutt commented:
**win2016-agent-bolt**

```
==> win2016-agent-bolt: Running provisioner: shell...
    win2016-agent-bolt: Running: tests/provision_basic_win.ps1 as c:\tmp\vagrant-shell.ps1
    win2016-agent-bolt: Installing Puppet from https://downloads.puppetlabs.com/windows/puppet5/puppet-agent-x64-latest.msi
    win2016-agent-bolt: Log will be written to C:/vagrant/puppet-agent.log
    win2016-agent-bolt: symbolic link created for C:\ProgramData\PuppetLabs\code\environments\production\modules\sensu <<===>> C:\vagrant
    win2016-agent-bolt: Out-File : Could not find a part of the path 'C:\ProgramData\PuppetLabs\puppet\etc\hiera.yaml'.
    win2016-agent-bolt: At C:\tmp\vagrant-shell.ps1:36 char:18
    win2016-agent-bolt: + $hiera_content | Out-File -FilePath "$hiera_file" -Encoding ascii
    win2016-agent-bolt: +                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : OpenError: (:) [Out-File], DirectoryNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : FileOpenFailure,Microsoft.PowerShell.Commands.OutFileCommand
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install puppetlabs-stdlib
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install puppetlabs-chocolatey
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install puppet-archive
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install puppet-windows_env
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install richardc-datacat
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: Copy-Item : Cannot find path 'C:\vagrant\tests\ssl' because it does not exist.
    win2016-agent-bolt: At C:\tmp\vagrant-shell.ps1:48 char:1
    win2016-agent-bolt: + Copy-Item -Path "C:\vagrant\tests\ssl\*" -Destination "C:\ProgramData ...
    win2016-agent-bolt: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (C:\vagrant\tests\ssl:String) [Copy-Item], ItemNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.CopyItemCommand
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet resource host sensu-backend.example.com ensure=present ip=192. ...
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
==> win2016-agent-bolt: Running provisioner: shell...
    win2016-agent-bolt: Running: tests/test_bolt_win.ps1 as c:\tmp\vagrant-shell.ps1
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet module install puppetlabs-chocolatey
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: puppet : The term 'puppet' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + puppet apply -e 'include ::chocolatey'
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (puppet:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: choco : The term 'choco' is not recognized as the name of a cmdlet, function, script file, or operable program. Check
    win2016-agent-bolt: the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + choco install -y puppet-bolt
    win2016-agent-bolt: + ~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (choco:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
    win2016-agent-bolt: bolt : The term 'bolt' is not recognized as the name of a cmdlet, function, script file, or operable program. Check
    win2016-agent-bolt: the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + bolt task run sensu::install_agent backend=sensu-backend.example.com: ...
    win2016-agent-bolt: + ~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (bolt:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
==> win2016-agent-bolt: Running provisioner: shell...
    win2016-agent-bolt: Running: inline PowerShell script
    win2016-agent-bolt: facter : The term 'facter' is not recognized as the name of a cmdlet, function, script file, or operable program.
    win2016-agent-bolt: Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
    win2016-agent-bolt: At line:1 char:1
    win2016-agent-bolt: + facter --custom-dir=C:\vagrant\lib\facter sensu_agent
    win2016-agent-bolt: + ~~~~~~
    win2016-agent-bolt:     + CategoryInfo          : ObjectNotFound: (facter:String) [], CommandNotFoundException
    win2016-agent-bolt:     + FullyQualifiedErrorId : CommandNotFoundException
```

treydock commented:
For Debian and Ubuntu, the issue is that we can support `apt` module as low as `5.0.1` but because we install a version of stdlib that isn't supported by `5.0.1` then we get issues.  What we should be doing is installing versions of `apt` and `stdlib` that matches the range we support in README or in metadata.json, not just a specific version, and let `puppet module` install the best fit versions using our supported ranges.  This is the best approach to avoid these frequent breakages that aren't real bugs with this module.  This is also the behavior taken with acceptance tests and best to be consistent.

Adding commit to fix this.

treydock commented:
I got the Windows 2008r2 error to go away by bypassing vbguest plugin but new errors are inability to download Puppet because Windows 2008r2 doesn't work with TLSv1.1 or TLSv1.2 and the supported TLS doesn't work for downloading Puppet MSI.  I tried using chocolately and same TLS/SSL problem.  I think the OS is just so old it's going to be a lot of work to make it interface with modern websites using modern SSL configs.  The only way I can forsee fixing Windows 2008r2 box is building my own box with newer powershell since all the existing boxes are very old and don't appear to have modern powershell.  I would recommend we remove the broken box from Vagrant file.

treydock commented:
Fixed Windows boxes by using chocolately to install Puppet, also disabling `sensuctl configure` in Windows as we don't test that with Appveyor either and was only breaking on Windows 2012 and really only necessary on backend.

treydock commented:
The WIndows 2008r2 is not fixed, just 2012 and 2016.  See previous comments about 2008r2.

ghoneycutt commented:
I updated the vagrant plugins and el8 is now working.

ghoneycutt commented:
debian 8 is still an issue. it's problem is not related to the version of the apt module.

ghoneycutt commented:
debian 8 is related to vbguest. Perhaps it is too old?

treydock commented:
I destroyed my copy, updated my box from `8.11.0` to `8.11.1` and now I have issues. The errors I see appear to be apt related, getting 404 errors accessing some packages when vbguest tries to install necessary packages to build drivers.  I downgrade back to `8.11.0` and the issue goes away.

I added a commit to pin debian8 box to `8.11.0`

treydock commented:
Fixed SSL issues in a different way, instead using same approach taken in acceptance tests and setting `certname` puppet setting to match the SSL cert the box should use which allows module defaults to work.  This required setting `api_host` to `fqdn` for backend boxes which maintains the old behavior where certname defaults to fqdn.

ghoneycutt commented on Vagrantfile:
```diff
@@ -166,28 +166,14 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
   config.vm.define "debian8-agent", autostart: false do |agent|
     agent.vm.box = "debian/jessie64"
+    agent.vm.box_version = '8.11.0'
```
Add a note and a link to this PR that 8.11.1 is broken, so we know why it is pinned and perhaps can unpin when 8.12 comes out or whenever the fix is ready.

treydock replied:
Comment added.

ghoneycutt commented on Vagrantfile:
```diff
@@ -166,28 +166,14 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     agent.vm.provision :shell, :inline => "facter --custom-dir=/vagrant/lib/facter sensu_agent"
   end
 
-  config.vm.define "win2008r2-agent", autostart: false do |agent|
```
getting rid of this platform?

treydock replied:
We can still support it just not in Vagrant, I could not get it to work. Once I got rid of issues with older powershell not working with vbguest, I ran into issues with older powershell not being able to download Puppet MSI or Chocolately installer due to not supporting modern TLS that all websites are now using.  The only only solution I could think of is to create a custom box of Windows 2008R2 that has newer powershell installed and just doesn't seem worth the effort when it doesn't affect this module's ability to support a platform.

Also I'm pretty sure Windows 2008R2 is EOL.  I can open a different pull request to remove support for that platform, but figured could leave as Sensu Go lists it as supported.

treydock replied:
https://www.microsoft.com/en-us/cloud-platform/windows-server-2008

EOL back in January.

ghoneycutt replied:
Agreed, lets just remove it from Vagrant. If it turns into a problem supporting it later, we can re-evaluate then.

ghoneycutt commented:
@treydock this is failing appveyor

treydock commented:
That's the transient issue with appveyor. Nothing in this pull request impacts anything outside vagrant and vagrant isn't tested by travis or appveyor. The error is missing files that sometimes happens, like puppet files we install during setup that sometimes happens but never could reproduce.
